lib-tui/GHCup/Brick/Widgets/Menus/CompileGHC.hs

```diff
@@ -0,0 +1,207 @@
+  where
+    initialState =
+      CompileGHCOptions
+        (Right "")
+        Nothing
+        Nothing
+        Nothing
+        Nothing
+        []
+        False
+        Nothing
+        Nothing
+        Nothing
+        Nothing
+    -- Brick's internal editor representation is [mempty].
+    emptyEditor i = T.null i || (i == "\n")
+    whenEmpty :: a -> (T.Text -> Either Menu.ErrorMessage a) -> T.Text -> Either Menu.ErrorMessage a
+    whenEmpty emptyval f i = if not (emptyEditor i) then f i else Right emptyval
+
+    bootstrapV :: T.Text -> Either Menu.ErrorMessage (Either Version FilePath)
+    bootstrapV i =
+      case not $ emptyEditor i of
+        True  ->
+          let readVersion = bimap (const "Not a valid version") Left (version (T.init i)) -- Brick adds \n at the end, hence T.init
+              readPath = do
+                mfilepath <- filepathV i
+                case mfilepath of
+                  Nothing -> Left "Invalid Empty value"
+                  Just f  -> Right (Right f)
+           in if T.any isPathSeparator i
+                then readPath
+                else readVersion
+        False -> Left "Invalid Empty value"
+
+    versionV :: T.Text -> Either Menu.ErrorMessage (Maybe [VersionPattern])
+    versionV = whenEmpty Nothing (bimap T.pack Just . OptParse.overWriteVersionParser . T.unpack . T.init)  -- Brick adds \n at the end, hence T.init
+
+    jobsV :: T.Text -> Either Menu.ErrorMessage (Maybe Int)
+    jobsV =
+      let parseInt = bimap (const "Invalid value. Must be an integer") Just . readEither @Int . T.unpack
+       in whenEmpty Nothing parseInt
+
+    patchesV :: T.Text -> Either Menu.ErrorMessage (Maybe (Either FilePath [URI]))
+    patchesV = whenEmpty Nothing readPatches
+      where
+        readUri :: T.Text -> Either String URI
+        readUri = first show . parseURI . UTF8.fromString . T.unpack
+        readPatches j =
+          let
+            x = (bimap T.unpack (fmap Left) $ filepathV j)
+            y = second (Just . Right) $ traverse readUri (T.split isSpace j)
+          in first T.pack $ x <|> y
+
+    filepathV :: T.Text -> Either Menu.ErrorMessage (Maybe FilePath)
+    filepathV i =
+      case not $ emptyEditor i of
+        True  -> absolutePathParser (T.unpack i)
+        False -> Right Nothing
+
+    absolutePathParser :: FilePath -> Either Menu.ErrorMessage (Maybe FilePath)
+    absolutePathParser f = case isValid f && isAbsolute f of
+                  True -> Right . Just . stripNewlineEnd . normalise $ f
+                  False -> Left "Please enter a valid absolute filepath."
+
+    additionalValidator :: T.Text -> Either Menu.ErrorMessage [T.Text]
+    additionalValidator = Right . T.split isSpace
+
+    systemV :: T.Text -> Either Menu.ErrorMessage (Maybe BuildSystem)
+    systemV = whenEmpty Nothing readSys
+      where
+        readSys i
+          | T.toLower i == "hadrian" = Right $ Just Hadrian
+          | T.toLower i == "make"    = Right $ Just Hadrian
```

Comment:
?

Reply:
Also, the "build system" input field doesn't work for me. It always says "Not a valid Build System".

Reply:
I guess you get the "not valid" because Trailing new line from Brick's editor. I'll add a generic "drop traling newline"  when validating, because I've seen this a few times already